### ISSUE TITLE
Array of tables output

### ISSUE THREAD

**Issue by 89z** [marzer/tomlplusplus#73]
If I have this input:

~~~toml
[[Winter]]
December = [
   { Sunday = 6 },
   { Monday = 7 }
]
January = [
   { Sunday = 6 },
   { Monday = 7 }
]
[[Spring]]
March = [
   { Sunday = 6 },
   { Monday = 7 }
]
April = [
   { Sunday = 6 },
   { Monday = 7 }
]
~~~

with this program:

~~~c
#include <fstream>
#include <iostream>
#include <toml.hpp>

int main() {
   auto config = toml::parse_file( "configuration.toml" );
   std::cout << config << std::endl;
   return 0;
}
~~~

I get this result:

~~~toml
[[Spring]]
April = [ { Sunday = 6 }, { Monday = 7 } ]
March = [ { Sunday = 6 }, { Monday = 7 } ]

[[Winter]]
December = [ { Sunday = 6 }, { Monday = 7 } ]
January = [ { Sunday = 6 }, { Monday = 7 } ]
~~~

is it possible to get something closer to the input?

**Comment by marzer** [comment]
No, the parser and serializer do not preserve formatting. The serializer instead emits the TOML in a terse, "best-fit" format, only adding newlines between tables or when it would otherwise have to wrap (e.g. when an array's elements reach some width threshold).

Order isn't preserved either since the TOML spec doesn't mandate that order is important.

**Comment by 89z** [comment]
Hm, I see that now! It looks like this input produces exactly the same output:

~~~
[[Winter]]
December = [
   { Sunday = 6 },
   { Monday = 7 },
   { Tuesday = 8 },
   { Wednesday = 9 },
   { Thursday = 10 },
   { Friday = 11 },
   { Saturday = 12 },
   { Sunday = 13 }
]
~~~

interesting. I guess its not an issue then. Is it possible to control the wrap
length?

**Comment by marzer** [comment]
Yeah, it's hardcoded. Would be fairly easy to expose it as an option though.

**Comment by 89z** [comment]
would you accept a PR? I am not great with C++ but I can try to work something up

**Comment by marzer** [comment]
I'd accept a PR, sure, just give [CONTRIBUTING](https://github.com/marzer/tomlplusplus/blob/master/CONTRIBUTING.md) a quick read first.

Though if you're not too confident in your C++ I'm also happy with design feedback here in this thread. Either is ok.

**Comment by 89z** [comment]
probably several ways to handle this, but I was thinking something simple like

~~~c
#ifdef TOML_WRAP
   inline constexpr size_t default_formatter_line_wrap = TOML_WRAP;
#else 		
   inline constexpr size_t default_formatter_line_wrap = 120_sz; 
#endif
~~~

**Comment by marzer** [comment]
If you want to go that route, have a look at the various user config bits in `toml_preprocessor.hpp`. Generally I use the form
```cpp
#ifndef TOML_FANCY_NEW_CONFIG
    #define TOML_FANCY_NEW_CONFIG the_default_value
#endif
```
So that it's always defined and I don't have to branch on it later.

Eventually I'd like to make this a runtime-configurable option in the formatter interface, though allowing a config to override the default line length is a good start. I'd suggest the name `TOML_DEFAULT_LINE_WRAP`.

**Comment by 89z** [comment]
Do you mean this?

https://github.com/marzer/tomlplusplus/blob/master/include/toml++/toml_preprocessor.h

I didnt see a `toml_preprocessor.hpp`. Otherwise I think I have enough to make up a PR, unless you have any other suggestions.

**Comment by marzer** [comment]
Oh, yup. That's the one. My crappy legacy choice to mix `.h` and `.hpp` strikes again.

**Comment by 89z** [comment]
ok I think Im done, but I cant get the single header to rebuild. I tried this:

~~~
PS C:\tomlplusplus\python> python generate_single_header.py
Traceback (most recent call last):
  File "C:\tomlplusplus\python\generate_single_header.py", line 9, in <module>
    import utils
ModuleNotFoundError: No module named 'utils'
~~~

then this:

~~~
PS C:\tomlplusplus\python> python utils.py generate_single_header.py
~~~

second command didnt give any error, but it also didnt change the single header file

**Comment by marzer** [comment]
Er. Wtf.

Try a normal CMD shell instead of PS?

**Comment by 89z** [comment]
Seem same result

**Comment by marzer** [comment]
What about `py generate_single_header.py`?

Also, which version of python is that on your machine? The scripts are written in python 3.

**Comment by 89z** [comment]
I think I see the issue. If I run this:

~~~
python utils.py generate_single_header.py
~~~

It just runs `utils.py` with `generate_single_header.py` as an argument.
`utils.py` is just a bunch of functions, so nothing happens. If I run this:

~~~
python generate_single_header.py
~~~

I get this:

~~~
Traceback (most recent call last):
  File "C:\tomlplusplus\python\generate_single_header.py", line 9, in <module>
    import utils
ModuleNotFoundError: No module named 'utils'
~~~

I think it has to do with this:

https://github.com/marzer/tomlplusplus/blob/9ea4438fc4ac7f19d158e80b05e0123090dda583/python/generate_single_header.py#L9

I dont think that is the right way to import from a local file:

https://stackoverflow.com/questions/4142151

**Comment by marzer** [comment]
Ah, I see. Odd that it's only been an issue for you now, but has worked for others. Welp, guess I have some python stuff to learn!

If your changes are trivially simple you can manually edit toml.hpp such that the changes are exactly mirrored there, and it should work OK.

**Comment by 89z** [comment]
correction, the Python stuff was my error:

https://stackoverflow.com/a/65226460

sorry about that. Will try to finish up PR now

**Comment by 89z** [comment]
I think I am doing something wrong. If I do a fresh clone of the repo, and then
use this file:

~~~
[[Winter]]
December = [
   { Sunday = 6 },
   { Monday = 7 },
   { Tuesday = 8 },
   { Wednesday = 9 },
   { Thursday = 10 },
   { Friday = 11 },
   { Saturday = 12 },
   { Sunday = 13 }
]
~~~

I get the same input and output, as expected. But if I go here:

https://github.com/marzer/tomlplusplus/blob/9ea4438fc4ac7f19d158e80b05e0123090dda583/toml.hpp#L8340

and just manually changed that line to `80_sz`, then I get this result:

~~~
[[Winter]]
December = [ { Sunday = 6 }, { Monday = 7 }, { Tuesday = 8 }, { Wednesday = 9 }, { Thursday = 10 }, { Friday = 11 }, { Saturday = 12 }, { Sunday = 13 } ]
~~~

**Comment by marzer** [comment]
Oh, yah, it stops doing multi-line expansion of objects once `indent + width >= default_formatter_line_wrap ` (see `impl::default_formatter_forces_multiline`). Doesn't make much sense here, though. I should revisit that logic.

**Comment by marzer** [comment]
Fixed in `54d80bb`.

**Comment by 89z** [comment]
https://github.com/marzer/tomlplusplus/pull/74
